### DISCUSSION TITLE
How to access the renovate analysis result programatically?

### DISCUSSION THREAD

**Discussion by karfau** [renovatebot/renovate#15621] [Request Help]
We would like to get access to the following information that seems to be used to render the Dependency Dashboard:

- How many update PR will there be (if possible by category in the dependency dashboard)
- Mow many major/minor/path update PR are there and for which packages

since we want to create a report per repository (without clicking around and manually counting things, and also enhanced by other data like github security information).

Is there any kind of API to access that data either from the logs (I think I spotted some JSON there?) or by posting the data to some webhook after each run?
Or would we need to parse the description of the Dependency Dashboard?

**Comment by viceice** [comment]
there is no API. so you need to parse logs or dashboard

**Reply by karfau** [reply]
Are you saying there is also no API to access the logs? Would we need to manually download the log file?
Or is there some cli tool that we could execute locally in a repository to get access to that data/those logs?

**Reply by astellingwerf** [reply]
You would need to capture the log file after it's been created, you can set the environment variables:

- RENOVATE_LOG_FILE=renovate-log.ndjson
- RENOVATE_LOG_FILE_LEVEL=debug

That file is then parsable as a Newline Delimited JSON log file

**Reply by viceice** [reply]
if you're using the hosted GitHub app, yes, there is no public API. for self-hosted renovate you can upload the logs to any host after run manually.

**Reply by karfau** [reply]
Thank you for the explanation, do you think it's worth the effort to do the "self-hosted" setup just to receive the logs, but stick to GitHub for doing the actual thing? (How much effort might it be?)

**Reply by fullammo** [reply]
I was able to get my required updates with the following search term in the log file: `flattened updates`.

Is there any way I could visualize or create an html output with the renovate tool of this "report" of sort with the actual and up to date versions located grouped by the "managers" (for eg. docker/node/terraform/npm)?

**Reply by secustor** [reply]
Take a look at [this talk](https://secustor.dev/talks/oss_2024_europe_extract_dependency_data_on_scale_with_renovate/) especially slides 30 and later.

**Reply by muffl0n** [reply]
I made an approach with parsing the log file with `jq` leveraging https://github.com/bats-core/bats-core to run the tests.

This is my approach:
tests.bats
```
###
# Run with "bats tests.bats --jobs $(grep -c "@test" tests.bats) -T"

setup_file() {
  export RENOVATE_LOGFILE=../log/renovate-log.ndjson
  if [ -f $RENOVATE_LOGFILE ]; then
    echo "File \"$RENOVATE_LOGFILE\" already exists, not running Renovate. To force a run of Renovate, delete this file." >&3
  else
    cd ..
    RENOVATE_REPOSITORIES="[\"mygroup/renovate-test\"]" RENOVATE_LOG_FILE_LEVEL=trace task
    cd - || exit
  fi
}

@test "check for ignorePaths" {
  jq '.ignorePaths | select(. != null).[] | test("^\\.teamcity/\\*\\*")' "$RENOVATE_LOGFILE" -e
}

@test "check for authentication in hostRules" {
  jq '. | select (.msg | test("^Adding password authentication for software.example.com \\(hostType=maven\\) to hostRules"))' "$RENOVATE_LOGFILE" -e
  jq '. | select (.msg | test("^Adding password authentication for docker.example.com \\(hostType=docker\\) to hostRules"))' "$RENOVATE_LOGFILE" -e
}

@test "check for various config values" {
  jq '. | select(.msg == "Combined config").config["pre-commit"].enabled' "$RENOVATE_LOGFILE" -e
  jq '. | select(.msg == "Combined config").config["gitLabIgnoreApprovals"]' "$RENOVATE_LOGFILE" -e
  jq '. | select(.msg == "Combined config").config["timezone"] | test("^Europe/Berlin$")' "$RENOVATE_LOGFILE" -e
  jq '. | select(.msg == "Combined config").config["prConcurrentLimit"] | select(. == 20)' "$RENOVATE_LOGFILE" -e
  jq '. | select(.msg == "Combined config").config["prHourlyLimit"] | select(. == 20)' "$RENOVATE_LOGFILE" -e
  jq '. | select(.msg == "Combined config").config["hashedBranchLength"] | select(. == 35)' "$RENOVATE_LOGFILE" -e
}

@test "check for opentofu registry" {
  jq '. | select(.msg == "Found preset local>mygroup/renovatebot:opentofu.json5").presetConfig.packageRules[] | select(.matchDatasources == ["terraform-provider", "terraform-module"]).registryUrls[] | select(. == "https://registry.opentofu.org")' "$RENOVATE_LOGFILE" -e
}

#############################################
# Check files to be parsed by managers
#############################################
@test "check for parsing of .tofu files with manager terraform" {
  jq '. | select(.msg == "Using file match: \\.tofu$ for manager terraform")' "$RENOVATE_LOGFILE" -e
}

@test "check for parsing of main.tofu and versions.tofu with manager terraform" {
  jq '. | select(.msg == "Matched 2 file(s) for manager terraform: main.tofu, versions.tofu")' "$RENOVATE_LOGFILE" -e
}

@test "check for parsing of .gitlab-ci-include.yml files with manager gitlabci-include" {
  jq '. | select(.msg == "Using file match: ^\\.gitlab\\-ci\\-include\\.yml$ for manager regex")' "$RENOVATE_LOGFILE" -e
}

#############################################
# Check for upgrades of packages
#############################################
@test "check for grouping of google and google-beta in versions.tofu" {
  jq '.branchesInformation | select(. != null).[] | select(.prTitle | startswith("Update Terraform google-provider to")).upgrades | length | select (. == 2)' "$RENOVATE_LOGFILE" -e

  jq '.branchesInformation | select(. != null).[] | select(.prTitle | startswith("Update Terraform google-provider to")).upgrades[0].depName | test("^google$")' "$RENOVATE_LOGFILE" -e
  jq '.branchesInformation | select(. != null).[] | select(.prTitle | startswith("Update Terraform google-provider to")).upgrades[0].packageFile | test("^versions.tofu$")' "$RENOVATE_LOGFILE" -e

  jq '.branchesInformation | select(. != null).[] | select(.prTitle | startswith("Update Terraform google-provider to")).upgrades[1].depName | test("^google-beta$")' "$RENOVATE_LOGFILE" -e
  jq '.branchesInformation | select(. != null).[] | select(.prTitle | startswith("Update Terraform google-provider to")).upgrades[1].packageFile | test("^versions.tofu$")' "$RENOVATE_LOGFILE" -e
}
...
```
When there is no logfile (`../log/renovate-log.ndjson`), bats runs `RENOVATE_REPOSITORIES="[\"mygroup/renovate-test\"]" RENOVATE_LOG_FILE_LEVEL=trace task` in its `setup_file` method. We have a `Taskfile.yaml` for this to ease running Renovate locally.

For some tests, like the check for `"Found preset local>mygroup/renovatebot:opentofu.json5"`, it's necessary to run Renovate with log level `trace`. This is kind of a bummer cause the logfile easily gets several hundreds of MB in size which significantly slows down parsing with `jq`. Running tests in parallel helps a bit.

I created https://github.com/renovatebot/renovate/pull/35011 which changes the loglevel for those outputs from `trace` to `debug`.

**Reply by viceice** [reply]
use log level remap instead

**Reply by muffl0n** [reply]
Works like a charm! Didn't know about that possibility. Thank you very much! <3
